### PR TITLE
Compile javascript assets with webpack.

### PR DESCRIPTION
**Pull request by jcoyne**
This will enable us to write javascript in es6+ and compile it into a
bundle for distribution.

I still want to see if we can require the client application to use webpacker to pull in blacklight as a javascript dependency rather than using sprockets.  This would cause the app to pull the dependencies like bootstrap too.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/projectblacklight/blacklight/issues"
+  },
+  "homepage": "https://github.com/projectblacklight/blacklight#readme",
+  "devDependencies": {
+    "webpack": "^3.3.0"
+  },
+  "dependencies": {
+    "bootstrap": "4.0.0-alpha.6",
```
Comment (cbeer) [inline package.json]: beta1?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "blacklight",
+  "version": "1.0.0",
```
Comment (cbeer) [inline package.json]: Ought this be versioned with blacklight?

Comment (jcoyne) [inline package.json]: Yes, that's a good idea.

File: package.json (hunk 1 above)
Comment (jcoyne) [inline package.json]: 👍

**Comment by mejackreed** [comment]
I'm generally 👎  on this PR unless Rails upstream moves to using Webpacker by default. Blacklight seems like it has traditionally followed Rails default conventions, and while its nice to use ES6 I don't think diverging is worth it yet. I think this will cause more adoption/upgrade discomfort than its worth. Its a new paradigm and it doesn't feel fully baked yet.

**Comment by jcoyne** [comment]
@mejackreed for the record this PR allows us to use webpacker, but does not require the upstream to use webpacker. You can continue to use sprockets if you want to.

File: README.md [hunk 3]
```diff
@@ -38,5 +38,81 @@ rails generate blacklight:install
...
+
+## Building the javascript
+The javascript is built by npm from sources in `app/javascript` into a bundle
+in `app/assets/javascripts/blacklight/blacklight.js`. This file should not be edited
+by hand as any changes would be overwritten.
+
+This is accomplished with the following steps:
+1. Install npm
+1. run `npm install` to download dependencies
+1. run `npm run-script js-compile-bundle` to build the bundle
```
[the remark this answers is not in the document]
Comment (jcoyne) [reply README.md]: This was copied directly out of bootstrap. https://github.com/twbs/bootstrap/blob/9057d8da164bed683f49d30c6414753dad41de3e/package.json#L40  My thoughts are that the `js-` could differentiate it from other bundling tasks (e.g. `css-`)

File: README.md [hunk 4]
```diff
@@ -38,5 +38,81 @@ rails generate blacklight:install
...
+
+## Building the javascript
+The javascript is built by npm from sources in `app/javascript` into a bundle
+in `app/assets/javascripts/blacklight/blacklight.js`. This file should not be edited
+by hand as any changes would be overwritten.
+
+This is accomplished with the following steps:
+1. Install npm
+1. run `npm install` to download dependencies
+1. run `npm run js-compile-bundle` to build the bundle
```
Comment (camillevilla) [inline README.md]: Running this command generates a JS file with double quotes and other Hound complaints. Should we have Hound ignore this file?

File: README.md [hunk 5]
```diff
@@ -38,5 +38,81 @@ rails generate blacklight:install
...
+1. run `npm publish` to push the javascript package to https://npmjs.org/package/blacklight-frontend
+
+## Using the javascript
+Install Webpacker
```
Comment (camillevilla) [inline README.md]: Either add a link to [Webpack's install page](https://webpack.js.org/guides/installation/) or include :
```
npm install --save-dev webpack
```

Comment (jcoyne) [reply README.md]: Webpacker rather than Webpack: https://github.com/rails/webpacker#installation

**Comment by jrochkind** [comment]
@cbeer suggested I put this comment here. @cbeer wrote:

> And the thought is... if you're loading the javascript individually, you probably have your own javascript preprocessor if you care about IE11? I guess I can buy that.

I replied:

I think maybe there is in fact no way to load the Javascript individually via sprockets even before this PR? That went away when the new system of supporting dual sprockets/npm was created?

I don't think I've been able to do it since then. I do miss it. I wish I could do it. It was a problem before this PR. This PR may not make it easier to step out of it, if it's desired to step out of it -- you saying something about it makes me think for the first time it was a bug/error/confusion on my part that i wasn't able to load the JS individually via sprockets, instead of an intended design? I always figured before now it was known and intended that this wasn't something Blacklight supported anymore.

I never liked it though! I just figured I had no choice, just had to deal with it.

It would be interesting to get clarification on whether Blacklight intends to support loading individual JS files via sprockets... I'm not sure there is anyone who can answer this question though?  If Blacklight does intend that, this PR definitely makes it harder...

**Pull request merged**